crates/wasmtime/src/module/serialization.rs

```diff
@@ -0,0 +1,715 @@
+        self.check_features(compiler)?;
+
+        // Check the flags last as they are the least helpful in terms of diagnostic message
+        self.check_flags(isa)?;
+
+        let types = self
+            .tables
+            .into_iter()
+            .map(|t| Arc::new(t.unwrap_owned()))
+            .collect::<Vec<_>>();
+        let module = mk(engine, &types, self.data)?;
+
+        // Validate the module can be used with the current allocator
+        engine.allocator().validate(module.inner.module.module())?;
+
+        return Ok(module);
+
+        fn mk(
+            engine: &Engine,
+            types: &Vec<Arc<TypeTables>>,
+            data: SerializedModuleData<'_>,
+        ) -> Result<Module> {
+            let mut artifacts = CompiledModule::from_artifacts_list(
+                data.artifacts
+                    .into_iter()
+                    .map(|i| i.unwrap_owned())
+                    .collect(),
+                engine.compiler().isa(),
+                &*engine.config().profiler,
+            )?;
+            let inner = ModuleInner {
+                engine: engine.clone(),
+                types: types[data.type_tables].clone(),
+                module: artifacts.pop().unwrap(),
+                artifact_upvars: artifacts,
+                module_upvars: data
+                    .modules
+                    .into_iter()
+                    .map(|m| mk(engine, types, m))
+                    .collect::<Result<Vec<_>>>()?,
+            };
+
+            Ok(Module {
+                inner: Arc::new(inner),
+            })
+        }
+    }
+
+    fn check_version(&self) -> Result<()> {
+        if self.version != env!("CARGO_PKG_VERSION") {
+            bail!(
+                "Module was compiled with Wasmtime version '{}'",
+                self.version
+            );
+        }
+
+        Ok(())
+    }
+
+    fn check_triple(&self, isa: &dyn TargetIsa) -> Result<()> {
+        let triple = target_lexicon::Triple::from_str(&self.target).map_err(|e| anyhow!(e))?;
+
+        if triple.architecture != isa.triple().architecture {
+            bail!(
+                "Module was compiled for architecture '{}'",
+                triple.architecture
+            );
+        }
+
+        if triple.operating_system != isa.triple().operating_system {
+            bail!(
+                "Module was compiled for operating system '{}'",
+                triple.operating_system
+            );
+        }
+
+        Ok(())
+    }
+
+    fn check_flags(&self, isa: &dyn TargetIsa) -> Result<()> {
+        let host_level = isa.flags().opt_level().into();
+        if self.opt_level != host_level {
+            bail!("Module was compiled with optimization level '{:?}' but '{:?}' is expected for the host", self.opt_level, host_level);
+        }
+
+        if self.flags_hash != Self::simple_hash(isa.flags()) {
+            bail!("Module was compiled with different Cranelift flags than the host");
+        }
```

Comment:
I always get this confused with the `isa_flags` below. Do you think we could use different names for these flags? IIRC there's like a two-level table of flags for cranelift ISAs, the "this applies to all targets" table and the "this applies to only one target" table. I think this is checking the target-independent table, right? And the `check_isa_flags` below is checking the target-specific table?

I basically found it a bit confusing to have this and `check_isa_flags` below.

Reply:
Agreed, it is confusing.  `flags_hash` here is the hash of the common/shared Cranelift flags and the `isa_flags` are those enabled by the target ISA (currently these are 1:1 with CPU features).

I can see about a better name.

Reply:
I've removed `flags_hash` in favor of storing all of the shared and ISA-specific settings in the compiled module.